File: mssp/archaeology/018-object-freeze/src/island_test.mjs
// The island test, run against the real built-ins.
//
//   node src/island_test.mjs
//
// Section 3 is the control. Section 3b is the drill: a mode that DECLARES it
// reports violations while swallowing them must be caught by running it.
import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

import * as upstream from "./SMS/upstream.mjs";
import * as sloppy from "./TMS/modes/sloppy.mjs";
import * as strict from "./TMS/modes/strict.mjs";

const here = path.dirname(fileURLToPath(import.meta.url));
const FMS = JSON.parse(fs.readFileSync(path.join(here, "FMS", "architecture.json"), "utf8"));
const POLICY = JSON.parse(fs.readFileSync(path.join(here, "SCL", "policy.json"), "utf8"));
const failures = [];
let ran = 0;
const check = (label, ok, detail = "") => {
  ran += 1;
  process.stdout.write(`  ${ok ? "PASS" : "FAIL"}  ${label}${detail ? ` - ${detail}` : ""}\n`);
  if (!ok) failures.push(label);
};
const say = (line = "") => process.stdout.write(`${line}\n`);

say(`\n  ${upstream.runtime()}`);

say("\n== 1. each mode is an island, and FMS matches the tree");
for (const [unit, declared] of Object.entries(FMS.units)) {
  const dir = path.join(here, ...unit.split("/"));
  const onDisk = fs.readdirSync(dir).filter((n) => n.endsWith(".mjs")).sort();
  check(`${unit}: FMS declares what is on disk`,
    JSON.stringify(onDisk) === JSON.stringify([...declared].sort()),
    `disk ${onDisk.join(", ")} | FMS ${[...declared].sort().join(", ")}`);
  for (const file of onDisk) {
    const body = fs.readFileSync(path.join(dir, file), "utf8");
    check(`${unit}/${file} imports nothing at all`, !/^\s*import\s/m.test(body));
  }
}
check("both modes declare whether they report violations",
  [sloppy, strict].every((m) => typeof m.REPORTS_VIOLATIONS === "boolean"));
check("and how they are entered", [sloppy, strict].every((m) => Boolean(m.HOW_IT_IS_ENTERED)));

say("\n== 2. the same assignment against the same frozen object");
const viaSloppy = upstream.through(sloppy, upstream.frozen());
const viaStrict = upstream.through(strict, upstream.frozen());
check("sloppy raised nothing", viaSloppy.threw === null);
check("and the object kept its old value", viaSloppy.actual === 100);
check("but the expression evaluated to the new one", viaSloppy.returned === 999);
check("so `returned` and `actual` disagree about the same write",
  viaSloppy.returned !== viaSloppy.actual, `${viaSloppy.returned} vs ${viaSloppy.actual}`);
check("strict raised a TypeError for the identical line", /^TypeError/.test(viaStrict.threw ?? ""),
  viaStrict.threw ?? "");
check("and its object kept the old value too", viaStrict.actual === 100);
check("so the object agrees across modes and only the report differs",
  viaSloppy.actual === viaStrict.actual && (viaSloppy.threw === null) !== (viaStrict.threw === null));

say("\n== 3. the control - the same assignment against an object never frozen");
const openSloppy = upstream.through(sloppy, upstream.neverFrozen());
const openStrict = upstream.through(strict, upstream.neverFrozen());
check("it raises nothing under either mode", openSloppy.threw === null && openStrict.threw === null);
check("the write lands", openSloppy.actual === 999 && openStrict.actual === 999);
check("and it also returns 999 - the control", openSloppy.returned === 999);
check("so `returned` is the same in the success and in the silent failure",
  openSloppy.returned === viaSloppy.returned,
  `${openSloppy.returned} both times, actual ${openSloppy.actual} vs ${viaSloppy.actual}`);
check("which is what makes `returned` uninformative rather than merely wrong",
  openSloppy.returned === viaSloppy.returned && openSloppy.actual !== viaSloppy.actual);

say("\n== 3b. DRILL - a mode that overclaims must be caught by running it");
const liar = {
  MODE: "drill-liar",
  REPORTS_VIOLATIONS: true, // the declaration
  assign(target, key, value) { // the implementation, which swallows
    try { const returned = (target[key] = value); return { returned, threw: null, actual: target[key] }; }
    catch { return { returned: null, threw: null, actual: target[key] }; }
  },
};
check("the drill unit declares it reports violations", liar.REPORTS_VIOLATIONS === true);
check("running it says otherwise", upstream.reportsViolations(liar) === false);
check("so the declaration is refused", upstream.reportsViolations(liar) !== liar.REPORTS_VIOLATIONS);
check("and the honest declarations hold under the same probe",
  upstream.reportsViolations(sloppy) === sloppy.REPORTS_VIOLATIONS
  && upstream.reportsViolations(strict) === strict.REPORTS_VIOLATIONS);

say("\n== 4. the declarer was never given a way to say what a violation means");
check("Object.freeze takes one argument", upstream.freezeArity() === 1, `${upstream.freezeArity()}`);
check("and isFrozen answers the same under either mode", Object.isFrozen(upstream.frozen()));
check("so the object cannot tell which consumer it reached",
  FMS.modes.sloppy.how_it_is_entered.includes("none of which the declarer chose"));
check("the two entrances are different constructs, not a setting",
  sloppy.HOW_IT_IS_ENTERED !== strict.HOW_IT_IS_ENTERED);

say("\n== 5. the declaration is shallower than it reads");
const nested = upstream.nested();
check("the outer object is frozen", Object.isFrozen(nested));
check("the inner one is not", !Object.isFrozen(nested.inner));
const nestedStrict = strict.assign(nested.inner, "price", 999);
check("so a strict-mode write to the inner object raises nothing", nestedStrict.threw === null);
check("and it lands", nested.inner.price === 999);
check("which is a failure of the declaration, not of either mode",
  nestedStrict.threw === null && viaStrict.threw !== null);

say("\n== 6. what this does not change");
check("freezing is right - the object did keep its value under both modes",
  viaSloppy.actual === 100 && viaStrict.actual === 100);
check("sloppy mode cannot be fixed retroactively - it is what unmarked code means",
  sloppy.HOW_IT_IS_ENTERED.includes("classic script"));
check("SCL names what it cannot reach", POLICY.what_this_cannot_reach.includes("Function constructor"));
check("and this deployment does refuse", POLICY.a_silent_violation_is === "fatal");

say("");
if (failures.length > 0) {
  say(`  ${failures.length} FAILED: ${failures.join(" | ")}`);
  process.exitCode = 1;
} else {
  say(`  ${ran} checks passed - every probe ran the real built-ins`);
}

File: mssp/archaeology/019-dict-get-sentinel/src/island_test.py
"""The island test, run against real dicts.

    python src/island_test.py

Section 3 is the control. Section 3b is the drill: a reader that DECLARES it
separates the two situations while calling plain get must be caught.
"""
import json
import pathlib
import re
import sys

from SMS import upstream
from TMS.readers import forgiving, subscript, with_sentinel

HERE = pathlib.Path(__file__).parent
FMS = json.loads((HERE / "FMS" / "architecture.json").read_text(encoding="utf-8"))
POLICY = json.loads((HERE / "SCL" / "policy.json").read_text(encoding="utf-8"))
FAILURES = []
RAN = [0]


def check(label, ok, detail=""):
    RAN[0] += 1
    print(f'  {"PASS" if ok else "FAIL"}  {label}{f" - {detail}" if detail else ""}')
    if not ok:
        FAILURES.append(label)


print(f"\n  {upstream.versions()}")

print("\n== 1. each reader is an island, and FMS matches the tree")
for unit, declared in FMS["units"].items():
    directory = HERE.joinpath(*unit.split("/"))
    on_disk = sorted(p.name for p in directory.glob("*.py") if p.name != "__init__.py")
    check(f"{unit}: FMS declares what is on disk", on_disk == sorted(declared),
          f'disk {", ".join(on_disk)} | FMS {", ".join(sorted(declared))}')
    for name in on_disk:
        body = directory.joinpath(name).read_text(encoding="utf-8")
        check(f"{unit}/{name} imports nothing at all",
              not re.search(r"^\s*(import|from)\s", body, re.M))
check("all three readers declare what they separate",
      all(isinstance(m.SEPARATES_MISSING_FROM_NONE, bool)
          for m in (forgiving, with_sentinel, subscript)))

print("\n== 2. two situations, one value")
on_none = forgiving.read(upstream.PRESENT_NONE, "a")
on_missing = forgiving.read(upstream.MISSING, "a")
check("d.get returns None for a key set to None", on_none["value"] is None)
check("and None for a key that is not there", on_missing["value"] is None)
check("the same value", on_none["value"] == on_missing["value"])
check("the same type", type(on_none["value"]) is type(on_missing["value"]))
check("and the same object", on_none["value"] is on_missing["value"])
check("neither raised", on_none["raised"] is None and on_missing["raised"] is None)

print("\n== 3. the control - a key that is present with a truthy value")
table = upstream.falsy_table()
control = [row for row in table if "control" in row["case"]]
check("the table carries a control", len(control) == 1)
check("and it comes out on the other branch", control[0]["falsy"] is False)
check("while every other case is falsy",
      all(row["falsy"] for row in table if "control" not in row["case"]))
collapsed = upstream.collapsed(table)
present = upstream.of_those_present(table)
check("so 6 situations reach one branch", collapsed == 6, f"{collapsed}")
check("and 5 of those 6 have the key", present == 5, f"{present}")
check("which makes the idiomatic test right about absence once in six",
      collapsed - present == 1)
check("without the control the sentence would be about the test, not the values",
      control[0]["falsy"] != table[0]["falsy"])

print("\n== 3b. DRILL - a reader that overclaims must be caught by running it")


class Liar:
    READER = "drill-liar"
    SEPARATES_MISSING_FROM_NONE = True  # the declaration

    @staticmethod
    def read(mapping, key):  # the implementation, which does not
        return {"value": mapping.get(key), "raised": None}


check("the drill unit declares it separates them", Liar.SEPARATES_MISSING_FROM_NONE is True)
check("running it says otherwise", upstream.separates(Liar) is False)
check("so the declaration is refused", upstream.separates(Liar) != Liar.SEPARATES_MISSING_FROM_NONE)
for module in (forgiving, with_sentinel, subscript):
    check(f"and {module.READER}'s declaration holds under the same probe",
          upstream.separates(module) is module.SEPARATES_MISSING_FROM_NONE)

print("\n== 4. the discriminators that exist")
check("`key in d` separates them",
      ("a" in upstream.PRESENT_NONE) != ("a" in upstream.MISSING))
sent_none = with_sentinel.read(upstream.PRESENT_NONE, "a")
sent_missing = with_sentinel.read(upstream.MISSING, "a")
check("d.get(key, SENTINEL) separates them", sent_none["present"] != sent_missing["present"])
check("and still returns the same value in both", sent_none["value"] == sent_missing["value"],
      "which is why the answer needs two fields, not a better single one")
sub_missing = subscript.read(upstream.MISSING, "a")
check("d[key] raises on the missing one", sub_missing["raised"] is not None, sub_missing["raised"])
check("and returns quietly on the present-None one",
      subscript.read(upstream.PRESENT_NONE, "a")["raised"] is None)
check("so the forgiving and the strict accessor live on the same type",
      forgiving.RAISES_ON_MISSING is False and subscript.RAISES_ON_MISSING is True)
check("the sentinel is not in the signature - the caller has to bring an object()",
      "default" in (dict.get.__doc__ or "") and with_sentinel.MISSING is not None)

print("\n== 5. what this does not change")
check("d.get is right to be forgiving - a default is the common case",
      forgiving.read({"a": 7}, "a")["value"] == 7)
check("and None is a legal value, so it cannot be the missing marker",
      upstream.PRESENT_NONE["a"] is None)
check("which is exactly why the sentinel has to be an object the caller owns",
      with_sentinel.MISSING is not None and with_sentinel.MISSING is not False)
check("SCL names what it cannot reach", "call site" in POLICY["what_this_cannot_reach"])
check("and this deployment does refuse the forgiving reader",
      POLICY["reader"] != forgiving.READER)

print()
if FAILURES:
    print(f'  {len(FAILURES)} FAILED: {" | ".join(FAILURES)}')
    sys.exit(1)
print(f"  {RAN[0]} checks passed - every probe ran real dicts")
